storage/src/main/java/com/epam/aidial/core/storage/service/ResourceService.java

```diff
@@ -278,24 +278,28 @@ public boolean hasResource(ResourceDescriptor descriptor) {
     }
 
     @Nullable
-    public Pair<ResourceItemMetadata, String> getResourceWithMetadata(ResourceDescriptor descriptor) {
-        return getResourceWithMetadata(descriptor, true);
+    public Pair<ResourceItemMetadata, String> getResourceWithMetadata(ResourceDescriptor descriptor, EtagHeader etag) {
+        return getResourceWithMetadata(descriptor, true, etag);
     }
 
     @Nullable
-    private Pair<ResourceItemMetadata, String> getResourceWithMetadata(ResourceDescriptor descriptor, boolean lock) {
+    private Pair<ResourceItemMetadata, String> getResourceWithMetadata(ResourceDescriptor descriptor, boolean lock, EtagHeader etagHeader) {
         String redisKey = redisKey(descriptor);
         Result result = redisGet(redisKey, true);
 
         if (result == null) {
             try (var ignore = lock ? lockService.lock(redisKey) : null) {
                 result = redisGet(redisKey, true);
-
+                String etag;
                 if (result == null) {
                     String blobKey = blobKey(descriptor);
                     result = blobGet(blobKey, true);
+                    etag = result.etag;
                     redisPut(redisKey, result);
+                } else {
+                    etag = result.etag;
                 }
+                etagHeader.validate(etag);
```

Comment:
This line checks etag only when a resource is not cached (loaded from blob), but skips any checks when it is in Redis.

Reply:
It should be check only when a result is present.
```
 if (result.exists()) {
    etagHeader.validate(result.etag());
    return Pair.of(...);
 }

```

Reply:
fixed